crates/core/component/community-pool/src/component/rpc.rs

```diff
@@ -0,0 +1,86 @@
+use penumbra_sdk_proto::core::{
+    asset::v1::{AssetId, Value},
+    component::community_pool::v1::{
+        query_service_server::QueryService, CommunityPoolAssetBalancesRequest,
+        CommunityPoolAssetBalancesResponse,
+    },
+};
+
+use async_stream::try_stream;
+use futures::{StreamExt, TryStreamExt};
+use std::pin::Pin;
+use tonic::Status;
+use tracing::instrument;
+
+use cnidarium::Storage;
+
+use super::StateReadExt;
+
+pub struct Server {
+    storage: Storage,
+}
+
+impl Server {
+    pub fn new(storage: Storage) -> Self {
+        Self { storage }
+    }
+}
+
+#[tonic::async_trait]
+impl QueryService for Server {
+    /// Stream of asset balance info within the CommunityPool, to satisfy the "repeated"
+    /// field in the protobuf spec.
+    type CommunityPoolAssetBalancesStream = Pin<
+        Box<
+            dyn futures::Stream<Item = Result<CommunityPoolAssetBalancesResponse, tonic::Status>>
+                + Send,
+        >,
+    >;
+
+    #[instrument(skip(self, request))]
+    async fn community_pool_asset_balances(
+        &self,
+        request: tonic::Request<CommunityPoolAssetBalancesRequest>,
+    ) -> Result<tonic::Response<Self::CommunityPoolAssetBalancesStream>, Status> {
+        let state = self.storage.latest_snapshot();
+        let request = request.into_inner();
+
+        // Asset IDs are optional in the req; if none set, return all balances.
+        let asset_ids: Vec<AssetId> = request.asset_ids;
```

Comment:
For correctness I think we to use the actual domain type here, because that's the one where we actually control the equality function, and that way we don't run into weird things if for some reason the RPC returns an alt denom or something.